core/lib/instrument.c

```diff
@@ -6328,6 +6328,146 @@ dr_insert_cbr_instrumentation_help(void *drcontext, instrlist_t *ilist, instr_t
 #elif defined(RISCV64)
     /* FIXME i#3544: Not implemented */
     ASSERT_NOT_IMPLEMENTED(false);
+#elif defined(AARCH64)
+    dcontext_t *dcontext = (dcontext_t *)drcontext;
+    ptr_uint_t address, target;
+    reg_id_t dir = DR_REG_NULL;
+    reg_id_t flags = DR_REG_NULL;
+    int opc;
+    ;
+    CLIENT_ASSERT(drcontext != NULL,
+                  "dr_insert_cbr_instrumentation: drcontext cannot be NULL");
+    address = (ptr_uint_t)instr_get_translation(instr);
+    CLIENT_ASSERT(address != 0,
+                  "dr_insert_cbr_instrumentation: can't determine app address");
+    CLIENT_ASSERT(instr_is_cbr(instr),
+                  "dr_insert_cbr_instrumentation must be applied to a cbr");
+    target = (ptr_uint_t)opnd_get_pc(instr_get_target(instr));
+
+    /* Compute branch direction */
+    opc = instr_get_opcode(instr);
+    if (opc == OP_cbnz || opc == OP_cbz) {
+        opnd_t reg_op = instr_get_src(instr, 1);
+        reg_id_t reg = opnd_get_reg(reg_op);
+        /* use dir register to compute direction */
+        dir = (reg == DR_REG_X0 || reg == DR_REG_W0) ? DR_REG_X1 : DR_REG_X0;
+        /* save old value of dir register to SPILL_SLOT_1 */
+        dr_save_reg(dcontext, ilist, instr, dir, SPILL_SLOT_1);
+        /* use flags register to save nzcv */
+        flags = (reg == DR_REG_X2 || reg == DR_REG_W2) ? DR_REG_X3 : DR_REG_X2;
+        /* save old value of flags register to SPILL_SLOT_2 */
+        dr_save_reg(dcontext, ilist, instr, flags, SPILL_SLOT_2);
+        /* save flags to flags register */
+        dr_save_arith_flags_to_reg(dcontext, ilist, instr, flags);
+
+        /* compare reg against zero */
+        instr_t *cmp = INSTR_CREATE_cmp(dcontext, reg_op, OPND_CREATE_INT(0));
+        MINSERT(ilist, instr, cmp);
+        /* compute branch direction */
+        opnd_t dir_op = opnd_create_reg(dir);
+        instr_t *cset = INSTR_CREATE_csinc(
+            dcontext, dir_op, OPND_CREATE_ZR(dir_op), OPND_CREATE_ZR(dir_op),
+            opnd_create_cond(opc == OP_cbnz ? DR_PRED_EQ : DR_PRED_NE));
+        MINSERT(ilist, instr, cset);
+    } else if (opc == OP_tbnz || opc == OP_tbz) {
+        opnd_t reg_op = instr_get_src(instr, 1);
+        reg_id_t reg = opnd_get_reg(reg_op);
+        reg_id_t dir_same_width = DR_REG_NULL;
+
+        /* use dir register to compute direction */
+        if (DR_REG_START_64 <= reg && reg <= DR_REG_STOP_64) {
+            /* 64-bit register */
+            dir = (reg == DR_REG_X0) ? DR_REG_X1 : DR_REG_X0;
+            dir_same_width = (reg == DR_REG_X0) ? DR_REG_X1 : DR_REG_X0;
+        } else {
+            /* 32-bit register */
+            dir = (reg == DR_REG_W0) ? DR_REG_X1 : DR_REG_X0;
+            dir_same_width = (reg == DR_REG_W0) ? DR_REG_W1 : DR_REG_W0;
```

Comment:
Looks like you can use API routines to make this simpler, replacing this if/else lines 6378-6386 with this:
```
dir = (reg_to_pointer_sized(reg) == DR_REG_X0) ? DR_REG_X1 : DR_REG_X0;
dir_same_width = reg_resize_to_opsz(dir, opnd_get_size(reg));
```